File: Results/DataReader.py
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import rosbag

# Read CSV files using pandas to handle headers
M = pd.read_csv('/home/vandalsnike/Results/trajectory_done.csv')
M2 = pd.read_csv('/home/vandalsnike/Results/trajectory.csv')

# Convert DataFrames to NumPy arrays
M = M.to_numpy()
M2 = M2.to_numpy()

# Plot the trajectories
plt.figure(1)
plt.plot(M[:, 1], M[:, 3], 'r', linewidth=2)
plt.axis('equal')
plt.plot(M2[:, 1], M2[:, 3], 'b-.', linewidth=2)
plt.title("Trajectory Data: Trefoil Knot in XZ Plane")
plt.scatter(M2[0, 1], M2[0, 3], color='m', marker='*')
plt.scatter(M2[-1, 1], M2[-1, 3], color='g', marker='*')
plt.legend(['Required', 'Executed', 'Starting point', 'Ending point'])
plt.xlabel("X")
plt.ylabel("Z")
plt.grid(True)

# Read rosbag
bag = rosbag.Bag("/home/vandalsnike/Results/joint_states_record_2024-09-05-22-13-46.bag")
joints_positions = []
joints_velocities = []

for topic, msg, t in bag.read_messages(topics=['/joint_states']):
    time = t.to_sec()
    positions = list(msg.position[:7])  # Extracting first 7 positions
    velocities = list(msg.velocity[:7])  # Extracting first 7 velocities
    joints_positions.append([time] + positions)
    joints_velocities.append([time] + velocities)

bag.close()

joints_positions = np.array(joints_positions)
joints_velocities = np.array(joints_velocities)

# Plot positions and velocities
plt.figure(2)
plt.subplot(1, 2, 1)
for i in range(1, 8):
    plt.scatter(joints_positions[:, 0], joints_positions[:, i], label=f'joint{i}', s=10)
plt.title("Joints' positions")
plt.legend()
plt.xlabel('Time')
plt.grid(True)

plt.subplot(1, 2, 2)
for i in range(1, 8):
    plt.plot(joints_velocities[:, 0], joints_velocities[:, i], label=f'joint{i}')
plt.title("Joints' velocity")
plt.legend()
plt.xlabel('Time')
plt.grid(True)

plt.tight_layout()
plt.show()

File: Results/DataReader.m
close all
clear all

% Taking last file from joint_data folder
fileList = dir("/home/vandalsnike/catkin_ws7/Results/JointData");
fileList = fileList(~[fileList.isdir]);
[~, idx] = max([fileList.datenum]);

% Reading data
M = readmatrix('trajectory.csv');
M2 = readmatrix('trajectory_done.csv');
bag=rosbag(fileList(idx).name);

% Component over time
figure(1)
sgtitle("Components Over Time")
subplot(4,3,[1 2 3])
plot(M(:,1),M(:,2),'r','LineWidth',1.4)
hold on
plot(M2(:,1),M2(:,2),'b-.','LineWidth',1.4)
legend('Required','Executed')
xlabel("Time (s)")
ylabel("X (m)")
subplot(4,3,[4 5 6])
plot(M(:,1),M(:,4),'r','LineWidth',1.4)
hold on
plot(M2(:,1),M2(:,4),'b-.','LineWidth',1.4)
legend('Required','Executed')
xlabel("Time (s)")
ylabel("Z (m)")
subplot(4,3,[7 8 9])
plot(M(:,1),M(:,3),'r','LineWidth',1.4)
hold on
plot(M2(:,1),M2(:,3),'b-.','LineWidth',1.4)
legend('Required','Executed')
xlabel("Time (s)")
ylabel("Y (m)")

quat_req = [M(:,8), M(:,5:7)];  
quat_exe = [M2(:,8), M2(:,5:7)];  

euler_req = zeros(size(M(:, 5:8), 1), 3);
euler_exe = zeros(size(M2(:, 5:8), 1), 3);

for i = 1:size(M(:, 5:8), 1)
    q = quat_req(i, :);
    euler_req(i, :) = rad2deg(quat2eul(q,'XYZ'));
end

for i = 1:size(M2(:, 5:8), 1)
    q2 = quat_exe(i, :);
    euler_exe(i, :) = rad2deg(quat2eul(q2,'XYZ'));
end

subplot(4,3,10);
plot(M(:, 1),  euler_req(:, 1), 'b', 'DisplayName', 'Required','LineWidth',1.4); 
hold on;
plot(M2(:, 1), abs(euler_exe(:, 1)), 'r--', 'DisplayName', 'Executed','LineWidth',1.2);
title('Roll over time');
xlabel('Time (s)'); 
ylabel('Roll (degree)');
legend('show', 'Location', 'southeast');
grid on;

subplot(4,3,11);
plot(M(:, 1), euler_req(:, 2), 'b', 'DisplayName', 'Required','LineWidth',1.4); 
hold on;
plot(M2(:, 1),euler_exe(:, 2), 'r--', 'DisplayName', 'Executed','LineWidth',1.2);
title('Pitch over time');
xlabel('Time (s)'); 
ylabel('Pitch (degree)');
legend('show', 'Location', 'southeast');
grid on;

subplot(4,3,12);
plot(M(:, 1), euler_req(:, 3), 'b', 'DisplayName', 'Required','LineWidth',1.4); 
hold on;
plot(M2(:, 1), euler_exe(:, 3), 'r--', 'DisplayName', 'Executed','LineWidth',1.2);
title('Yaw over time');
xlabel('Time (s)'); 
ylabel('Yaw (degree)');
legend('show', 'Location', 'southeast');
grid on;




% Trajectory on XZ plane
figure(2)
plot(M(:,2),M(:,4),'r','LineWidth',1.4)
axis equal
hold on
plot(M2(:,2),M2(:,4),'b-.','LineWidth',0.9)
title("Trajectory Data: Trefoil Knot in XZ Plane")
scatter(M2(1,2),M2(1,4),'m*')
scatter(M2(end,2),M2(end,4),'g*')
legend('Required','Executed','Start point','End point','Location','best')
axis equal
xlabel("X (m)")
ylabel("Z (m)")



% Reading message from .bag file
msgs = readMessages(bag);

joints_velocities=zeros(size(msgs,1),8);
joints_positions=zeros(size(msgs,1),8);

for i=1:length(msgs)
    time=bag.MessageList{i,1};
    positions=msgs{i,1}.Position(1:7);
    velocities=msgs{i,1}.Velocity(1:7);

    joints_positions(i,:)=[time positions'];
    joints_velocities(i,:)=[time velocities'];

end


% Joint data over time
figure(3)
subplot(2,1,1)
plot(joints_positions(:,1), joints_positions(:,2:end), 'LineWidth', 1.3)
title("Joint position")
legend('joint1','joint2','joint3','joint4','joint5','joint6','joint7')
xlabel('Time (s)')
hold on
smoothed_velocities = movmean(joints_velocities(:,2:end), 200);
subplot(2,1,2)
plot(joints_velocities(:,1), smoothed_velocities, 'LineWidth', 1.3)  
title("Joint velocity (smoothed)")
legend('joint1','joint2','joint3','joint4','joint5','joint6','joint7')
xlabel('Time (s)')


% Trajectory on 3D space
figure
plot3(M(:,2),M(:,3),M(:,4),'r')
hold on
plot3(M2(:,2),M2(:,3),M2(:,4),'b--')

